feat: gsoc phase 1 report of jaydip (#218)
CircuitVerse/Blog: content/posts/Jaydip_Dey_Phase_1_Report.md

  - **Loading Progress:**
    - On the simulator page, if `question_id` is found in localStorage, the corresponding circuit data is loaded
  
 ![add-question-simulator](/images/Jaydip_GSoC24/add-q-sim.webp) 

  - **Saving Data:**
    - When the "Save" button on the simulator page is clicked, the circuit data is saved to localStorage

**Add question flowchart**
![add-question-flowchart](/images/Jaydip_GSoC24/add-q-flow.webp)

#### Dropdowns for Category and Difficulty Level

- **Category:**
  - Fetched from the database (e.g., logic gates, combinational circuit, sequential circuit).
- **Difficulty Level:**
  - Enum values (easy, medium, hard, expert).

**Add question preview**
{{< video src="/videos/Jaydip_GSoC24/Add-question.mov" type="video/mp4" preload="auto" >}}

For more detailed information, please refer to [this](https://medium.com/@jaydipdey2807/week-5-6-at-circuitverse-google-summer-of-code-2024-87ea6f738cb4) blog post

---

### Pull Request

- PR : [Creating controllers, schemas and tables for circuitverse practice section](https://github.com/CircuitVerse/CircuitVerse/pull/4967)

> - PR : [Addig moderators and questions with simulator integration](https://github.com/CircuitVerse/CircuitVerse/pull/5015)

Reviewer: ```suggestion
- PR : [Adding moderators and questions with simulator integration](https://github.com/CircuitVerse/CircuitVerse/pull/5015)
```

date: 2024-07-10
draft: false
author: Jaydip Dey
type: post
---

This blog is dedicated to summarise the progress for the first phase of the project "CircuitVerse Practice Section" in Google Summer of Code'2024.

![phase-1](/images/Jaydip_GSoC24/Phase-1.png)

## About the Project

The project aims to address the need for a comprehensive practice section within CircuitVerse, enhancing learning and engagement in digital logic design. It includes developing a Question Bank Management system for convenient access to categorized questions, each featuring a Circuit Template with input/output probes and a pre-configured testbench. Markdown support will be integrated for improved readability and flexibility in displaying questions. A Question Bank Moderator role will be introduced for selected users to add and modify questions. Answers submitted by users will be automatically verified using CircuitVerse's testbench feature, enhancing the learning experience. Additionally, a Progress Dashboard within user profiles will track submission history and progress, with users able to choose between public or private visibility.

In the first phase, we have covered
* Community Bonding Period

> * Creating controllers, schemas and tables for all the use case [Week 1 to 3]

Reviewer 2: ```suggestion
* Creating controllers, schemas and tables for circuitverse practice section [Week 1 to 3]
```

* Manage question moderators[Week 4]

> * Moderator Question Creation and Integration[Week 5 to 6]

Reviewer 2: Integration with Simulator ?

---

### Community Bonding Period

> During the community bonding period for GSoC at CircuitVerse, we engaged in various activities to prepare for the project. We began by participating in meet-and-greet sessions with mentors, contributors, and admins, where we shared our excitement for GSoC, discussed our journeys, and introduced our projects. Mentors [Vaibhav Upreti](https://www.linkedin.com/in/vaibhav-upreti/), [Vedant Jain](https://www.linkedin.com/in/vedant-jain-781006145/), and [Tanmoy Sarkar](https://www.linkedin.com/in/tanmoysrt/), answered questions about the CircuitVerse Practice Section project, and [Aboobacker MK](https://www.linkedin.com/in/aboobacker-m-k/) provided insights into the 12-week project process and discussed open defects in CircuitVerse.

Reviewer 2: ```suggestion
During the community bonding period for GSoC at CircuitVerse, there were various activities to provide more context for the project. The project begins with a meet-and-greet sessions with mentors, contributors, and admins, where everyone shared their excitement for GSoC, discussed their journeys, and discuessed about the projects. Mentors [Vaibhav Upreti](https://www.linkedin.com/in/vaibhav-upreti/), [Vedant Jain](https://www.linkedin.com/in/vedant-jain-781006145/), and [Tanmoy Sarkar](https://www.linkedin.com/in/tanmoysrt/), answered questions about the CircuitVerse Practice Section project, and [Aboobacker MK](https://www.linkedin.com/in/aboobacker-m-k/) provided insights into the 12-week project process and discussed open defects in CircuitVerse.
```

You can add a context about `open defects in CircuitVerse`

Following this, we revisited key technologies by refreshing knowledge of Ruby on Rails through The Odin Project documentation and Vue JS with a YouTube crash course. This preparation was essential for upcoming work on the CircuitVerse codebase.

> We then focused on a detailed exploration of the CircuitVerse codebase, identifying areas for modification and making initial changes. we had in-depth discussions with [Ruturaj Mohite](https://www.linkedin.com/in/ruturaj-mohite/), about potential edge cases and the development of the question practice simulator versus the standard simulator, including how to track and save user progress.

Reviewer 2: ```suggestion
Then, the focus got shifted to detailed exploration of the CircuitVerse codebase, identifying areas for modification and making initial changes. we had in-depth discussions with [Ruturaj Mohite](https://www.linkedin.com/in/ruturaj-mohite/), about potential edge cases and the development of the question practice simulator versus the standard simulator, including how to track and save user progress.
```

      <button type="button" class="btn btn-secondary ml-0" data-toggle="modal" data-target="#addModeratorsModal"><%= t("users.circuitverse.index.add_moderator") %></button>
    <% end %>

    <%= render partial: "add_moderators_modal" %>
    ```

- **Modal for Managing Moderators:**
  - When the button is clicked, a modal opens where admins can enter email addresses of users to be added as moderators or remove existing moderators.

![manage-moderator-1](/images/Jaydip_GSoC24/manage_mod_1.webp)

#### Controller Functions

- **Add Moderators:**
  - The `add_moderators` function processes the email addresses entered by the admin and updates the `question_bank_moderator` field for the corresponding users.

    ```
    def add_moderators
      emails = params[:emails].split(',').map(&:strip)
      users = User.where(email: emails)
      
      users.each do |user|
        user.update(question_bank_moderator: true)
      end
      
      if users.all? { |user| user.errors.empty? }
        redirect_back(fallback_location: root_path)
      else
        render json: { errors: users.map(&:errors).flat_map(&:full_messages) }, status: :unprocessable_entity
      end
    end

>     ```

Reviewer 2: Avoid adding code snippets to this blog, until unless it's too important.
ig All sort of these things are present in your personal blog. Adding link to that blog or code snippet will be better way.

Reviewer 2: Same for all the code snippets in the blog @jaydip1235

Reviewer 2: You may add gif or video of the feature in the blog also to showcase a ppreview of the feature.

          const urlParams = new URLSearchParams(window.location.search);
          const questionId = urlParams.get('question_id');
       
          if (questionId) {
              let fl = 1;
              const data = generateSaveData("Untitled", fl);
              const localStorageKey = `${questionId}`;
              localStorage.setItem(localStorageKey, data);
              alert("Circuit boilerplate and test data saved");
          } else {
              alert("Error: Something went wrong! Try again");
          }
      });
      ```
**Add question flowchart**
![add-question-flowchart](/images/Jaydip_GSoC24/add-q-flow.webp)

#### Dropdowns for Category and Difficulty Level

- **Category:**
  - Fetched from the database (e.g., logic gates, combinational circuit, sequential circuit).
- **Difficulty Level:**
  - Enum values (easy, medium, hard, expert).

**Add question preview**
{{< video src="/videos/Jaydip_GSoC24/Add-question.mov" type="video/mp4" preload="auto" >}}

For more detailed information, please refer to [this](https://medium.com/@jaydipdey2807/week-5-6-at-circuitverse-google-summer-of-code-2024-87ea6f738cb4) blog post

---

### Pull Request

> Pull request for phase 1 can be found [here](https://github.com/CircuitVerse/CircuitVerse/pull/5015)

Reviewer 2: This PR https://github.com/CircuitVerse/CircuitVerse/pull/4967 should be added in the list also.

---

### Future Work

> The tasks to be done include developing the UI for the moderators dashboard and integrating it with the backend to enable CRUD operations on questions. Next, efforts will focus on making the question practice page responsive across all screens, implementing filtering logic and backend integration, and setting up auto-verification and submission of questions. Additionally, a progressive dashboard for users will be created to display submission history and progress, followed by backend integration and addition of test cases. 

Reviewer 2: It will be better if you can write this things in bullet points.
